### PR TITLE
Add unique constraints to prevent race conditions

### PR DESCRIPTION
**Pull request by jingcheng16**
## Technical Summary

Ticket: https://dimagi-dev.atlassian.net/browse/SAAS-14852
Add database constraints to `Invoice` and `CustomerInvoice` model
I didn't add it to `WireInvoice` model because wire invoice is manually triggered. It's not included in our periodic `generate_invoices` task.
Because we have many existing duplicates, we have a preliminary PR to [add `duplicate_invoice_id` column to the Invoice model](https://github.com/dimagi/commcare-hq/pull/33448). In this PR, we only add constraints to Invoice model when duplicate_invoice_id is null, (and `is_hidden_to_ops` is False).

### Safety story

Tested locally
Already take care of existing duplicates

### Automated test coverage

For a test that would catch it, are we looking for a test that have race condition and trigger the `IntegrityError` ?

### QA Plan

No

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by millerdev** [comment]
I'm not sure what impact the index creation will have on the database, but our usual practice when adding indexes is to create them `CONCURRENTLY` so they do not block other database operations while the index is being created. You can look at https://github.com/dimagi/commcare-hq/pull/24392 as an example of how that has been done in the past. Since this is creating a unique index, the SQL syntax will be `CREATE UNIQUE INDEX CONCURRENTLY ...`. The SQL that Django would run when applying the migration can be reviewed with this management command:
```sh
# forward SQL
./manage.py sqlmigrate accounting 0083

# reverse SQL
./manage.py sqlmigrate accounting 0083 --backwards
```

> are we looking for a test that have race condition and trigger the `IntegrityError`?

I think it would be useful to have a test that verifies that it is not possible to create duplicate invoices. The duplicate invoices in the test need not be created by the specific race condition that motivated this PR.

File: corehq/apps/accounting/models.py [hunk 1]
```diff
@@ -2131,6 +2131,13 @@ class Invoice(InvoiceBase):
 
     class Meta(object):
         app_label = 'accounting'
+        constraints = [
+            models.UniqueConstraint(
+                fields=['subscription', 'date_start', 'date_end'],
+                name='unique_invoice_per_subscription_period',
+                condition=models.Q(is_hidden_to_ops=False)
```
Comment (millerdev) [inline corehq/apps/accounting/models.py:2138]: Should there be a `duplicate_invoice_id IS NULL` condition too?
```suggestion
                condition=models.Q(is_hidden_to_ops=False, duplicate_invoice_id__isnull=True)
```

Comment (jingcheng16) [reply corehq/apps/accounting/models.py:2138]: Hi Daniel, the latest commit is just a merge from master. I haven't pushed my changes yet. :joy: I'll ping you when it's ready for review. I'll change the PR to draft now.

**Comment by jingcheng16** [comment]
@millerdev 
Hi Daniel, the PR is now ready to review.
But the test will pass, even if I roll back the migration. So, the test didn't hit the race condition I want to test. In this case, do we still want this test?

**Comment by millerdev** [comment]
I [added a commit](https://github.com/dimagi/commcare-hq/pull/33403/commits/bc6776a2df4d916629a2908c9d5ceb8e332b0761) to this PR that adds and uses a new manager function called `create_or_get`, which attempts the `create` first. It is more efficient for invoice creation because most of the time the invoice does not exist so the first `get` in `get_or_create` almost always fails.

The test now fails if the unique index does not exist. I also added a check on the number of invoices to verify that there are no duplicates.

**Comment by jingcheng16** [comment]
@millerdev 
Thank you for making this change! I think in this way, we are mainly rely on database constraints, but not the `get_or_create`

I only have one concern for this change:


`create_or_get` will be able to create invoice with same subscription, same start date, same end date, when `is_hidden_to_ops=True`. From the code I've seen so far, the invoice is always generated in `is_hidden_to_ops=False`, then at some point, when ops want to suppress the invoice, this operation will set `is_hidden_to_ops=True`
One condition in our db constraints is, `is_hidden_to_ops=False`. We do allow duplicate invoices when `is_hidden_to_ops=True`. So db constraints won't prevent this create.

With background knowledge, I know we won't create invoice with same subscription, same start date, same end date, when `is_hidden_to_ops=True`. But do we want to allow this possibility in the code? What if in the future, we start calling create invoice with `is_hidden_to_ops = True`? (Though very unlikely to happen...)

If I reverted your change, then I still have the question, do we actually need this test? Because the test is not testing the changes I made. 

I don't have any better ideas. I slightly prefer revert your commit and go without a test.

**Comment by millerdev** [comment]
I think it's important to test that the new constraint prevents duplicate invoices when the system should not allow them.

However, you make a good point about the new possibility of creating duplicate invoices with `is_hidden_to_ops=True`. Note that the race condition still exists if we use `get_or_create` in that scenario. I wonder if it would be better to eliminate the `is_hidden_to_ops` condition from the index to prevent that? In other words, make it so `duplicate_invoice_id` is required if you want to create a duplicate invoice, and all other duplicates are not allowed. I'd be happy to go to voice to discuss this more if that would be helpful.

**Comment by jingcheng16** [comment]
@millerdev 
Hi Daniel, I just (reverted your commit locally and) tested locally and our db constraints did work: 
I used to be able to reproduce the duplicate invoice issue by opening two shells, and hitting `generate_invoice` as quickly as possible. After having the changes, I now will get the expected error when I try to reproduce. 

With local testing, do you feel we can say our db constraints prevented duplicate invoices with confidence? 

I'm hesitated in changing the logic of existing invoice generation code, because I don't know if this will introduce some other problems that I cannot think of right now.

**Comment by millerdev** [comment]
An automated test would be preferable. Otherwise we would have to re-do your local test any time we touched related code, which seems unsustainable.

**Comment by biyeun** [comment]
@millerdev do you have any suggestions for how to trigger this race condition in an automated test? I tried finding examples of this in existing tests (say with `gevent`) and not seeing anything. I think Jing was having some difficulty with this.

I see that there is already a test for ensuring that the context surrounding the unique constraint functions as desired—is that not enough for this?

**Comment by millerdev** [comment]
@biyeun I think the problem with that test is that it does not fail if the unique constraint is removed, so it is currently not an effective test, or at least it was not effective before https://github.com/dimagi/commcare-hq/pull/33403/commits/bc6776a2df4d916629a2908c9d5ceb8e332b0761 was added to this PR. Jing had requested to remove it, but I think it would be better to make it an effective test than to remove it, so have been trying to find a way to accomplish that.

I do not think using a concurrency framework such as gevent will be a good way to trigger the race condition because I think that is likely to result in a flaky test. Instead, I think it's better to try to logically remove the race condition from the code paths that may trigger it. Using [`create_or_get` rather than `get_or_create`](https://github.com/dimagi/commcare-hq/pull/33403/commits/bc6776a2df4d916629a2908c9d5ceb8e332b0761) does that, but Jing is uncomfortable with that change. We talked today about a slightly different approach (use `get_or_create` when `is_hidden_to_ops` is true, else use `create_or_get`), which I believe she may be looking into now.

**Review by millerdev — APPROVED** [review]

**Pull request merged**